frame/assistants/shadow-liketh/src/mmr.rs

```diff
@@ -0,0 +1,76 @@
+use ckb_merkle_mountain_range::helper;
+
+// copy from https://github.com/darwinia-network/shadow/pull/165
+
+type Hash = [u8; 32];
+
+pub fn merge(lhs: &[u8], rhs: &[u8]) -> Hash {
+    let mut data = vec![];
+    data.append(&mut lhs.to_vec());
+    data.append(&mut rhs.to_vec());
+    let mut dest = [0; 32];
+    dest.copy_from_slice(blake2_rfc::blake2b::blake2b(32, &[], &data).as_bytes());
+    dest
+}
+
+pub fn bag_rhs_peaks(mut peaks: Vec<Hash>) -> ckb_merkle_mountain_range::Result<Hash> {
+    let last = peaks
+        .pop()
+        .ok_or(ckb_merkle_mountain_range::Error::GenProofForInvalidLeaves)?;
+    peaks.reverse();
+    Ok(peaks.iter().fold(last, |prev, &next| merge(&prev, &next)))
+}
+
+// mmr root
+// 1. helper::get_peaks(leaf_pos);
+// 2. query hash of peaks from db
+// 3. bag_rhs_peaks
+
+// first call gen_proof_positions to get the positions proof needed
+pub fn gen_proof_positions(mut pos: u64, leaf_pos: u64) -> (Vec<u64>, Vec<u64>, u64) {
```

Comment:
The implementation copy from https://github.com/darwinia-network/shadow/pull/165 is buggy.

In the test, the `proof_index `/ `merkle_proof_part` is empty. which indicate it is wrong.

We should use a more stable and production battle tested lib: https://github.com/nervosnetwork/merkle-mountain-range

Reply:
the `merkle_proof_part` is empty because the testcase use the same block_number as leaf_index to verify the number.
just like `./target/release/mmr https://api.thegraph.com/subgraphs/name/darwinia-network/ropsten-mmr 12244903 12244902`
if we use the next leaf_index to verify the block, it must not be empty, like
`./target/release/mmr https://api.thegraph.com/subgraphs/name/darwinia-network/ropsten-mmr 12244904 12244902`